pkg/kubelet/cm/deviceplugin/manager.go

```diff
@@ -461,166 +427,86 @@ func (m *ManagerImpl) readCheckpoint() error {
 	return nil
 }
 
-// updateAllocatedDevices gets a list of active pods and then frees any Devices that are bound to
-// terminated pods. Returns error on failure.
-func (m *ManagerImpl) updateAllocatedDevices(activePods []*v1.Pod) {
-	if !m.sourcesReady.AllReady() {
-		return
+// Allocate allocates a set of device resources for the Pod given.
+// TODO: invoke endpoint handler to lock it during device allocation.
+func (m *ManagerImpl) Allocate(node *schedulercache.NodeInfo, attrs *lifecycle.PodAdmitAttributes) error {
+	// Purge allocations by active Pods
+	if m.sourcesReady.AllReady() {
+		activePods := sets.NewString()
+		for _, p := range m.activePods() {
+			activePods.Insert(string(p.UID))
+		}
+		m.allocation.purge(activePods)
 	}
-	m.mutex.Lock()
-	defer m.mutex.Unlock()
-	activePodUids := sets.NewString()
-	for _, pod := range activePods {
-		activePodUids.Insert(string(pod.UID))
+
+	pod := attrs.Pod
+	var err error
+	request, _ := m.getPodRequests(pod)
+	selected, err := m.allocation.reserve(m.allDevices, request)
```

Comment:
I think there is a race condition here if we just pass allDevices to allocation without holding the manager lock.

Reply:
That was already mentioned higher :)

Reply:
Will revise. If I'm understanding this correctly, the key is to have manager process device allocation based on a snapshot of the known devices. There won't be such a solution that the manager can handle device changes during allocation because such asynchronous events are unpredictable. A device plugin may fail at any time. The best bet we have is to make the manager behave consistently.